driver-core/src/test/unit/com/mongodb/AuthConnectionStringTest.java

```diff
@@ -93,31 +92,65 @@ private void testValidUris() {
         }
 
         MongoCredential credential = connectionString.getCredential();
-        assertString("auth.db", credential.getSource());
-        assertString("auth.username", credential.getUserName());
+        if (credential != null) {
+            assertString("credential.source", credential.getSource());
+            assertString("credential.username", credential.getUserName());
+            assertMechanismProperties(credential);
+
+            // Passwords for certain auth mechanisms are ignored.
+            String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
+            if (password != null) {
+                assertString("credential.password", password);
+            }
+
+            assertMechanism("credential.mechanism", credential.getMechanism());
+        }
+    }
 
-        // Passwords for certain auth mechanisms are ignored.
-        String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
-        if (password != null) {
-            assertString("auth.password", password);
+    private void assertString(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isNull()) {
+            assertTrue(String.format("%s should be null", key), actual == null);
+        } else if (expected.isString()) {
+            String expectedString = expected.asString().getValue();
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expectedString), actual.equals(expectedString));
+        } else {
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expected), false);
         }
-        if (definition.get("options").isDocument()) {
-            for (Map.Entry<String, BsonValue> option : definition.getDocument("options").entrySet()) {
-                if (option.getKey().equals("authmechanism")) {
-                    String expected = option.getValue().asString().getValue();
-                    if (expected.equals("MONGODB-CR")) {
-                        assertNotNull(connectionString.getCredential());
-                        assertNull(connectionString.getCredential().getAuthenticationMechanism());
+    }
+
+    private void assertMechanism(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isString() && expected.asString().getValue().equals("MONGODB-CR")) {
+            assertTrue(String.format("%s should be null when the expected mechanism is MONGODB-CR", key), actual == null);
```

Comment:
Prefer `assertNull` to `assertTrue(val == null)`

Reply:
Done